### PR TITLE
new package: shadow

### PR DESCRIPTION
**Pull request by vsoch**
This will add shadow-uidmap, hoping to address #12297 @odoublewen I didn't hit the issue with semanage.h - does this mean I have it on my host (and still need to install here?)

I'm not sure that I did the configure and install steps correctly - please review! 

Signed-off-by: Vanessa Sochat <vsochat@stanford.edu>

### CODE REVIEW

**Comment by hartzell** [comment]
You might want to edit the title and note that the new package is actually named `shadow`.

**Comment by vsoch** [comment]
At least on the Singularity list it's referred to as shadow-uidmap, so in that there are other "shadow" things out there, it might be good to leave some tiny record of the uidmap bit.

File: var/spack/repos/builtin/packages/shadow/package.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+       user namespaces."""
+
+    homepage = "https://github.com/shadow-maint/shadow"
+    url      = "https://github.com/shadow-maint/shadow/releases/download/4.7/shadow-4.7.tar.gz"
+    git      = "https://github.com/shadow-maint/shadow.git"
+
+    version('4.7', sha256='5135b0ca2a361a218fab59e63d9c1720d2a8fc1faa520c819a654b638017286f')
+    version('4.6', sha256='4668f99bd087399c4a586084dc3b046b75f560720d83e92fd23bf7a89dda4d31')
+
+    version('v4.1.2.1', branch='v4.1.2.1')
```
Comment (hartzell) [inline var/spack/repos/builtin/packages/shadow/package.py]: Is there an actual need for these older versions?  They don't build for me (CentOS 7).  The devs didn't upload release tarballs, they only have the archive links.

Comment (vsoch) [reply var/spack/repos/builtin/packages/shadow/package.py]: I was just adding them to be complete, since the branches exist. I can definitely remove them.

Comment (vsoch) [reply var/spack/repos/builtin/packages/shadow/package.py]: Removed!

**Comment by odoublewen** [comment]
Good getting it working!  Huzzah, teamwork.  But what happened to the dependencies and the configure steps?  All I see in this PR now is the first 18 lines, ending in `version('4.6'...`

**Comment by vsoch** [comment]
@hartzell posted in the issue that we didn’t need it, so I removed it.

**Comment by vsoch** [comment]
Yeah so I tested again - I nuked everything in opt
```
$ rm -rf opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/
```
And installed:

```bash
$ bin/spack install shadow
==> Installing shadow
==> Searching for binary cache of shadow
==> Warning: No Spack mirrors are currently configured
==> No binary for shadow found: installing from source
==> Using cached archive: /tmp/spack/var/spack/cache/shadow/shadow-4.7.tar.gz
==> Staging archive: /tmp/vanessa/spack-stage/shadow-4.7-r76jmmomn7xmyumsjc4fj4fnbcgpjmqm/shadow-4.7.tar.gz
==> Created stage in /tmp/vanessa/spack-stage/shadow-4.7-r76jmmomn7xmyumsjc4fj4fnbcgpjmqm
==> No patches needed for shadow
==> Building shadow [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed shadow
  Fetch: 0.01s.  Build: 17.34s.  Total: 17.36s.
[+] /tmp/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/shadow-4.7-r76jmmomn7xmyumsjc4fj4fnbcgpjmqm
(base) vanessa@vanessa-ThinkPad-T460s:/tmp/spack$ ls opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/shadow-4.7-r76jmmomn7xmyumsjc4fj4fnbcgpjmqm/
bin  etc  sbin  share
```
And all seems well! So maybe using the different .tar.gz included any required dependencies? Is this weird / am I missing something?

**Comment by vsoch** [comment]
It's an `AutotoolsPackage` so the standard autogen, configure, build, install still applies (see print out).

**Comment by hartzell** [comment]
> And all seems well! So maybe using the different .tar.gz included any required dependencies? Is this weird / am I missing something?

GitHub has a concept of a release which is more than just a tag in the corresponding git repo; it's a thing that you create via the GUI/API and that has notes and various files attached.  If you don't explicitly create a release, GitHub will automagically create an impoverished version if/when it sees a tag w/ the right format (e.g. vMaj.Min.Patch).

GitHub download links that include `archive` are tarballs/zipfiles that are automagically generated on demand from the git repo at the corresponding tag.  They're exactly (+/-) what you'd get checking out the tree at that point.  You can think of that as the Raw Source, in this case the input to the various Autotools abomina^H^H^H^H^H^Hprograms.   GitHub download links that include `download` are release artifacts that the project maintainers have uploaded for that release (in the GithHub meaning of the term).

In this case the "download" tarballs have already been autotool'ed and *etc* and are ready to pop into the oven and build.

Looks like at v4.4 the shadow folk (scifi novel time!) started creating GitHub Releases(tm) that included various assets, including the source that you're using for the newer version.

When you were grabbing the `archive` versions, you were getting the Raw Source and you needed to figure out all of the magic and dependencies and *etc* that the authors used to convert that Raw Source into something useful.  Those dependencies were only needed for that step (proof by example...).

A really good example of this is Modern Perl packages, e.g. that use Dist::Zilla to build their releases.  The various `.pm` files in the source tree aren't ready to be released as is, they're passed through various tools that insert boilerplate docs (e.g. the LICENSE stanza), sort the various definitions into the One True Order, and *etc*, then packaged up into a tarball.  In most of the Perl sense of the word, that tarball (which would be uploaded to CPAN) is the "release".

**Comment by vsoch** [comment]
Holy crap, how does GitHub continue to amaze me! <3

**Comment by hartzell** [comment]
Re the name, [looks like the [corresponding Debian package](https://pkgs.org/download/uidmap) is just named `uidmap`, and I'm not trivially finding any CentOS/RHEL packages for it.  The obvious point of confusion would be the various shadow-utils packages out there for maintaining shadow password files.

I'm mostly just channelling the Spack lessons I've been taught.  Others (@adamjstewart , @scheibelp ) would have weightier votes.

**Comment by vsoch** [comment]
@hartzell you are a wealth of wisdom and resources yourself! If you say shadow, it shall be shadow! :sparkles:

**Comment by odoublewen** [comment]
I also learned something today.

I confirm that this builds in a fresh clone on a Centos7 box.  The executables appear appear happy, though I guess the pudding's proof will be how it works with singularity's fakeroot option.  

I'm going to PR an update to singularity that names `shadow` as a dep.  Currently testing it.  (I see they've already made a new release, 3.4, so I'll add that too.)

**Comment by hartzell** [comment]
> so the standard autogen, configure, build, install still applies (see print out).

For better or worse, the *autoreconf* step of the autotools package *always* runs, even when there's nothing to do.  Some might argue that its confusing that it runs when it's not actually doing anything.  Details [here](https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/autotools.py#L176-L180).

**Comment by vsoch** [comment]
woohoo! @odoublewen ping me when you do and I'd be happy to test it out. I started dinner so it might be tomorrow :P

**Comment by hartzell** [comment]
> a wealth of wisdom and resources yourself

And yet, there's [always more to learn](https://github.com/golang/go/issues/34091#issuecomment-528107672)....

**Pull request merged**